### PR TITLE
[INTEL_MKL] Enable MatMul + Bias + LeakyRelu fusion

### PR DESCRIPTION
**Pull request by retonym**
This PR enables  MatMul + Bias + LeakyRelu fusion with both MKL and Eigen implementation

### CODE REVIEW

**Comment by retonym** [comment]
Merge conflict is resolved. Please review this PR. Thanks

File: tensorflow/core/grappler/optimizers/remapper.cc [hunk 1]
```diff
@@ -471,8 +471,10 @@ bool FindContractionWithBiasAndActivation(
   // Currently, only matmul + bias + tanh is enable
   if (!IsMatMul(*contraction_node_def) && IsTanh(*node_def)) return false;
 
-  // Currently, only conv + bias + leakyrelu is enabled
-  if (!IsConv2D(*contraction_node_def) && IsLeakyRelu(*node_def)) return false;
+  // Currently, only (conv | matmul) + bias + leakyrelu is enabled
+  if ((!IsConv2D(*contraction_node_def) && !IsMatMul(*contraction_node_def)) &&
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper.cc:475]: Nit: This is slightly easier to read and is more consistent with the comment.
```suggestion
  if (!(IsConv2D(*contraction_node_def) || IsMatMul(*contraction_node_def)) &&
```

File: tensorflow/core/grappler/optimizers/remapper_test.cc [hunk 2]
```diff
@@ -697,6 +708,11 @@ class RemapperFuseMatMulWithBiasAndActivationTest : public RemapperTest {
           ASSERT_EQ(fused_ops.size(), 2);
           EXPECT_EQ(fused_ops[0], "BiasAdd");
           EXPECT_EQ(fused_ops[1], activation);
+
+          if (activation == "LeakyRelu") {
+            EXPECT_EQ(node.attr().at("leakyrelu_alpha").f(), leakyrelu_alpha);
+          }
+
```
Comment (penpornk) [inline tensorflow/core/grappler/optimizers/remapper_test.cc:715]: Nit: Maybe remove this blank line?

File: tensorflow/core/kernels/matmul_op_fused.cc [hunk 3]
```diff
@@ -102,6 +107,10 @@ struct LaunchFusedMatMulOp<CPUDevice, T> {
       case FusedComputationType::kBiasAddWithElu:
         executeWithOutputKernel(WithBiasAddAndElu<T>(bias_add_args));
         break;
+      case FusedComputationType::kBiasAddWithLeakyRelu:
+        out.device(d) = lhs.contract(rhs, dim_pair,
+                                     WithBiasAddAndLeakyRelu<T>(bias_add_args));
```
Comment (penpornk) [inline tensorflow/core/kernels/matmul_op_fused.cc:112]: Can we use `executeWithOutputKernel` for this?

File: tensorflow/core/ops/math_ops.cc [hunk 4]
```diff
@@ -954,6 +954,8 @@ REGISTER_OP("_FusedMatMul")
     .Attr("fused_ops: list(string) = []")
     // Attributes for the FusedBatchNorm ----------- //
     .Attr("epsilon: float = 0.0001")
+    // Attributes for the LeakyRelu ----------------------------------------- //
```
Comment (penpornk) [inline tensorflow/core/ops/math_ops.cc:957]: Nit: Maybe match the length with the comment lines around it?
```suggestion
    // Attributes for the LeakyRelu ---------------- //
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR and I'm sorry for the delay! I have a few minor comments.

File: tensorflow/core/grappler/optimizers/remapper.cc (hunk 1 above)
Comment (retonym) [reply tensorflow/core/grappler/optimizers/remapper.cc]: Thanks for the advice.  Code is changed.

File: tensorflow/core/grappler/optimizers/remapper_test.cc (hunk 2 above)
Comment (retonym) [reply tensorflow/core/grappler/optimizers/remapper_test.cc]: This blank line is deleted

File: tensorflow/core/kernels/matmul_op_fused.cc (hunk 3 above)
Comment (retonym) [reply tensorflow/core/kernels/matmul_op_fused.cc]: Yeah, the wrapper function is used

File: tensorflow/core/ops/math_ops.cc (hunk 4 above)
Comment (retonym) [reply tensorflow/core/ops/math_ops.cc]: Yeah, keep the code same length

**Comment by retonym** [comment]
Thanks for the review. The code is changed based on comments.

**Review by penpornk — COMMENTED** [review]
Thank you for the fixes! The PR looks good to me now.
@ezhulenev Could you please help take a look if the grappler parts are okay? I'll defer the approval to you.

**Review by ezhulenev — APPROVED** [review]

**Review by penpornk — APPROVED** [review]
Thank you, Eugene! 
(I'm approving too just so the PR status doesn't keep saying "changes requested".)

**Review by penpornk — CHANGES_REQUESTED** [review]
@retonym Could you please look into the remapper test failures in [Ubuntu CPU](https://source.cloud.google.com/results/invocations/888d1774-3f10-4bff-ad50-c56ab23cc302/targets/%2F%2Ftensorflow%2Fcore%2Fgrappler%2Foptimizers:remapper_test/log) and [Linux GPU](https://source.cloud.google.com/results/invocations/aadca1f2-b63a-45c1-99d5-f4ef158a40ef/targets/%2F%2Ftensorflow%2Fcore%2Fgrappler%2Foptimizers:remapper_test_gpu/log) CIs? Thank you!
```
[ RUN      ] RemapperFuseMatMulWithBiasAndActivationTest.F32
tensorflow/core/grappler/optimizers/remapper_test.cc:699: Failure
Expected equality of these values:
  node.op()
    Which is: "Tanh"
  "_FusedMatMul"
tensorflow/core/grappler/optimizers/remapper_test.cc:700: Failure
Expected: (node.input_size()) >= (3), actual: 1 vs 3
[  FAILED  ] RemapperFuseMatMulWithBiasAndActivationTest.F32 (11 ms)
```

**Comment by retonym** [comment]
The UT failure is fixed.  
MatMul+Bias+Tanh fusion is enabled in MKL only currently. It should not be tested in RemapperTest with normal Tensorflow.   
Merry Christmas.

**Review by penpornk — APPROVED** [review]
Thank you for the quick fix! Merry Christmas and happy holidays to you too! :)

**Pull request merged**